File: src/core/python/core/model/timetable.py
from typing import Callable


class Timetable(dict):
    """
    Mapping from events (of whatever type) to long integers,
    additionally maintaining a value for "time units per minute".
    """

    def __init__(self, timeUnitsPerMinute: int):
        """
        Create a new timetable with the given "time units per minute" value.
        :param timeUnitsPerMinute: time units per minute (may be fractional)
        """
        super(Timetable, self).__init__()
        self.timeUnitsPerMinute = timeUnitsPerMinute

    def getTimeUnitsPerMinute(self) -> int:
        """
        Returns the current "time units per minute" value for this timetable.
        :return: time units per minute (may be fractional)
        """
        return self.timeUnitsPerMinute

    def setTimeUnitsPerMinute(self, newTimeUnitsPerMinute: int,
                              roundingFunction:
                              Callable[[float], int]) -> None:
        """
        Updates the "time units per minute" value and converts existing
        timetable entries.
        :param new_timeUnitsPerMinute: new value for "time units per minute"
        :param roundingFunction: a rounding method used for all recalculations
        (Double to Long)
        """
        if newTimeUnitsPerMinute == self.timeUnitsPerMinute:
            return
        factor = 1.0 * newTimeUnitsPerMinute / self.timeUnitsPerMinute
        for event, time in self.items():
            self.__setitem__(event, roundingFunction(time * factor))

    def __str__(self) -> str:
        return ("Timetable: \n" + "\n".join("{}: {}".format(event, time)
                                            for event, time in self.items()))

File: src/core/java/src/net/lintim/model/Timetable.java
package net.lintim.model;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * Mapping from events (of whatever type) to long integers,
 * additionally maintaining a value for "time units per minute".
 */
public class Timetable<E> extends HashMap<E, Long> {

    private double timeUnitsPerMinute;

    /**
     * Create a new timetable with the given "time units per minute" value.
     * @param timeUnitsPerMinute - time units per minute (may be fractional)
     */
    public Timetable(double timeUnitsPerMinute) {
        super();
        this.timeUnitsPerMinute = timeUnitsPerMinute;
    }

    /**
     * Returns the current "time units per minute" value for this timetable
     * @return - time units per minute (may be fractional)
     */
    public double getTimeUnitsPerMinute() {
        return timeUnitsPerMinute;
    }

    /**
     * Updates the "time units per minute" value and converts existing timetable entries.
     * @param timeUnitsPerMinute - new value for "time units per minute"
     * @param roundingFunction - a rounding method used for all recalculations (Double to Long)
     */
    public void setTimeUnitsPerMinute(int timeUnitsPerMinute, Function<Double, Long> roundingFunction) {
        if (this.timeUnitsPerMinute == timeUnitsPerMinute) return;
        double factor = 1.0 * timeUnitsPerMinute / this.timeUnitsPerMinute;
        this.replaceAll((E key, Long oldValue) -> roundingFunction.apply(oldValue * factor));
        this.timeUnitsPerMinute = timeUnitsPerMinute;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Timetable: \n");
        for(Map.Entry<E, Long> entry : entrySet()){
            builder.append(entry.getKey()).append(":").append(entry.getValue());
        }
        return builder.toString();
    }
}
